### ISSUE TITLE
Class::disk(): Return value must be of type Illuminate\Filesystem\FilesystemAdapter, Sentry\Laravel\Features\Storage\SentryCloudFilesystem returned

### ISSUE THREAD

**Issue by Gummibeer** [getsentry/sentry-laravel#745]
### How do you use Sentry?

Sentry SaaS (sentry.io)

### SDK version

3.7.0

### Steps to reproduce

Have a method type-hinting for `Illuminate\Filesystem\FilesystemAdapter` and return `Storage:disk('s3')` or any other cloud disk.

### Expected result

Type-hint before v3.7.0 still works as expected and a breaking change of the package i sonly introduced during a new major-release.

### Actual result

TypeError is thrown with message:
> Return value must be of type `Illuminate\Filesystem\FilesystemAdapter`, `Sentry\Laravel\Features\Storage\SentryCloudFilesystem` returned

**Comment by atypic-garrett** [comment]
Came across the same issue, had to update composer to use 3.6.* :(

**Comment by Gummibeer** [comment]
There isn't anything helpful in the stacktrace - as the problem is the new sentry storage breadcrumbs and so on. And the rebind of the storages already in place.
Because of that the type of the resolved driver changes and this destroys all existing type-hints. To resolve that you will have to go back to v3.6 or do the work and type-hint the underlying interface instead of the precise class.

**Comment by cleptric** [comment]
In general, please assume that someone might not have all the necessary context to reproduce an issue immediately. In this case, a stack trace or your list of installed dependencies would have helped me verify that someone needs to install `league/flysystem-aws-s3-v3` to reproduce the issue.

I'm seeing the problem now, and we'll get this fixed momentarily 🙂

**Comment by Gummibeer** [comment]
It's not related to S3 specifically.  It's a problem with every single fileystem and related to your `Sentry/Laravel/Features/Storage/Integration.php:40` logic which rebinds every single fileystem with a new precise class. Because of that every single place with a precise class type-hint will now fail.

**Comment by stayallive** [comment]
@Gummibeer I understand this is not expected and we are looking for a solution other than just simply disabling the feature. 

But why are you type hinting the implementing class and not the interface (`Illuminate\Contracts\Filesystem\(Cloud)Filesystem`)?

**Comment by stayallive** [comment]
So in summary, this issue occurs when type-hinting the `Illuminate\Filesystem\FilesystemAdapter` instead of the `Illuminate\Contracts\Filesystem\Filesystem` or `Illuminate\Contracts\Filesystem\Cloud` contract. If you are affected these are the things you can do (preferably in this order):

First: See if you can adjust your type hint to use the contract, if you can type hint the contract that will solve the problem and no other changes are needed.

Second: You can disable the storage integration by updating the `config/sentry.php` config file you have published or by adding the settings to your `.env` if you do not have the config published:

In case you have an existing `config/sentry.php`, ensure you add the `storage` key to the `breacrumbs` and `tracing` config:

```php
'breadcrumbs' => [
    // Capture storage access as breadcrumbs
    'storage' => false,
],

'tracing' => [
    // Capture storage access as spans
    'storage' => false,
],
```

If you do not have it published you can add the following to your `.env`:

```
SENTRY_BREADCRUMBS_STORAGE_ENABLED=false
SENTRY_TRACE_STORAGE_ENABLED=false
```

Third: If all else fails you can update your `composer.json` from `"sentry/sentry-laravel": "^3.0"` (or similar) to `"sentry/sentry-laravel": "3.6.*"` for the time being locking you to the previous release that did not include the storage integration.

---

We are sorry for the trouble this has caused and are looking into if there is a way for us to keep the feature without introducing similar problems, for most people this will not be a problem but (as we've seen) it can be and we want to make sure we don't cause problems for anyone. Please bear with us while we work on coming up with a solution!

**Comment by Gummibeer** [comment]
@stayallive as it's userland code and if I bind that explicit class to that storage driver I can type-hint that class as I have registered it as the storage driver. And some precise implementations provide more features than the default interface.

If a package intercepts and alters all my bindings it should be something well documented in that package and not added during a minor release. As an OSS package I should typehint the contract - sure. But in userland I can become as precise as I want/need as I know what class is bound to the container and that shouldn't change without me upgrading a package's major version or me changing the bindings.

For me it was a relatively quick fix as I can work with the contract and was a quick search and replace in my app code. But it fucked my whole app for several minutes because of that `TypeError`.

**Comment by stayallive** [comment]
Thanks for the additional details @Gummibeer, please know we are human too and we made a mistake here and are working on preventing this happening in the future, please do let's keep it PG if possible, no reason to swear. I hope you can understand our situation too and that this was never our intention to break anyone's app for any reason and the change was made in good faith thinking we had covered all our bases.

**Comment by Gummibeer** [comment]
@stayallive all fine - could be that I sound rougher then I want to, it's 9pm here. 🙈

So in theory to be fully compatible with any bound storage driver you should use dynamic anonymous classes that will extend the given driver. Or at least document that storage drivers will be wrapped in these sentry classes and will lose the original driver class types.
Probably with a simple way (trait) to create my own `SentryS3FilesystemAdapter` or `SentryDropboxFilesystemAdapter` similar in which I can redirect potential additional calls and also keep my original type-hints as I can extend that `S3FilesystemAdapter` or whatever.

**Comment by stayallive** [comment]
All good suggestions, we are in the same timezone so I understand this is not the time for troubleshooting and nobody wants to deal with these things (at any time). We are picking this up first thing tomorrow and will take all your suggestions into consideration!

Originally we had this feature as a "manual" enable by setting a `sentry` driver (replacing the original driver) allowing us to wrap it that way (moving the original driver to another config key obviously so we know which one to wrap). 

However the automatic way also seemed to be equally safe and workable so we opted for less user friction and easier adoption, obviously not forseeing the type-hint issues. So we might revert to this being a documented feature instead of an out-of-the-box feature. Not sure if the anonymous class option is feasable, but this is more a braindump for tomorrow than anything 😄

**Comment by alquhaitsalla** [comment]
The same issue happened to us too, had to downgrade to 3.6

**Comment by bradleybernard** [comment]
+1 also getting this as well on production. Turned off the `storage` tracing and breadcrumbs for now, highlighted by Alex above. 

FWIW: My use case is using S3 and also Laravel FFmpeg, which takes in a storage driver in `fromDisk` method. See docs here if you're curious! https://github.com/protonemedia/laravel-ffmpeg

```php
use ProtoneMedia\LaravelFFMpeg\Support\FFMpeg;

$media = FFMpeg::fromDisk($diskName)->open($speechFilePath);
$duration = number_format($media->getDurationInMiliseconds() / 1000, 2);
```

It fails on the `fromDisk` method, exception:
```
TypeError: ProtoneMedia\LaravelFFMpeg\Filesystem\Disk::getFilesystemAdapter(): Return value must be of type Illuminate\Filesystem\FilesystemAdapter, Sentry\Laravel\Features\Storage\SentryCloudFilesystem returned
File "/app/blurred.php"
```

I think tracing storage calls is actually pretty sweet! Would love to turn this back on when it's fixed :) thanks for all the hard work!

**Comment by Gummibeer** [comment]
@bradleybernard this sounds like a bug @pascalbaljet could fix in the @protonemedia Laravel Media package. As it's a OSS package and I'm pretty sure that typing the underlying interface would be enough.

**Comment by mstaack** [comment]
having the same issues, downgrading sentry/sentry-laravel (3.7.0 => 3.6.1) helped for now

**Comment by pascalbaljet** [comment]
> @bradleybernard this sounds like a bug @pascalbaljet could fix in the @protonemedia Laravel ffmpeg package. As it's a OSS package and I'm pretty sure that typing the underlying interface would be enough.

I'm not fond of this change too much. I specifically target the `FilesystemAdapter` class instead of the `Filesystem` interface because of some missing methods on the `Filesystem` interface. I think the Sentry package automagically messing around with the `filesystems` config is more horrifying than me not targeting the interface.

**Comment by stayallive** [comment]
We agree, we are moving the feature to a manual one, it will be a documentation only feature you need to change your filesystem config for to enable, with the added warnings about what the problems could be with enabling. This because there is no bulletproof way for us to do this without catching some edge case even if we extend the `FilesystemAdapter`. See #746.

**Comment by spawnia** [comment]
https://github.com/getsentry/sentry-laravel/pull/746 looks fine as a quick fix, but I think we should still look into:
- simplifying the setup as much as possible, even when the feature is opt-in
- extending the `Illuminate\Filesystem\FilesystemAdapter` class instead of implementing its interfaces

**Comment by cleptric** [comment]
We just released version [3.7.1](https://github.com/getsentry/sentry-laravel/releases/tag/3.7.1) that addresses this issue.

We do apologise for the trouble this has caused!

**Comment by spawnia** [comment]
Sorry @stayallive and @cleptric and everyone affected for this issue. I readjusted my assumptions and propose https://github.com/getsentry/sentry-laravel/pull/750 to improve the ergonomics of this feature in a way that works better with the expectation that drivers will be specific classes.
